From 4fc163060a6cea5b853f04a4317cbba42ff1fcdb Mon Sep 17 00:00:00 2001
From: Sergey Fedorov <barracuda@macos-powerpc.org>
Date: Mon, 27 Jul 2026 20:36:46 +0000
Subject: [PATCH 13/16] Implement xdg-activation-v1

Vendor the staging xdg-activation-v1 protocol. Tokens are minted as
"owl-<counter>" strings on commit and never validated on activate, so
any token (including a client's own guess) is accepted; that's lenient
but sufficient for a v1 implementation, and matches foot's only use of
this protocol (bell.urgent). activate resolves the target wl_surface's
role to an OwlXdgToplevel, reaching its OwlWindow through the new
-windowWrapper accessor, and brings it forward with
requestUserAttention: when the app isn't active.
---
 Sources/Protocol/xdg-activation-v1.xml | 200 +++++++++++++++++++++++
 Sources/Server/OwlServer.m             |   2 +
 Sources/Shell/OwlXdgActivationV1.h     |  35 ++++
 Sources/Shell/OwlXdgActivationV1.m     | 211 +++++++++++++++++++++++++
 Sources/Shell/OwlXdgToplevel.h         |   2 +
 Sources/Shell/OwlXdgToplevel.m         |   4 +
 6 files changed, 454 insertions(+)
 create mode 100644 Sources/Protocol/xdg-activation-v1.xml
 create mode 100644 Sources/Shell/OwlXdgActivationV1.h
 create mode 100644 Sources/Shell/OwlXdgActivationV1.m

diff --git a/Sources/Protocol/xdg-activation-v1.xml b/Sources/Protocol/xdg-activation-v1.xml
new file mode 100644
index 0000000..9adcc27
--- /dev/null
+++ b/Sources/Protocol/xdg-activation-v1.xml
@@ -0,0 +1,200 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<protocol name="xdg_activation_v1">
+
+  <copyright>
+    Copyright © 2020 Aleix Pol Gonzalez &lt;aleixpol@kde.org&gt;
+    Copyright © 2020 Carlos Garnacho &lt;carlosg@gnome.org&gt;
+
+    Permission is hereby granted, free of charge, to any person obtaining a
+    copy of this software and associated documentation files (the "Software"),
+    to deal in the Software without restriction, including without limitation
+    the rights to use, copy, modify, merge, publish, distribute, sublicense,
+    and/or sell copies of the Software, and to permit persons to whom the
+    Software is furnished to do so, subject to the following conditions:
+
+    The above copyright notice and this permission notice (including the next
+    paragraph) shall be included in all copies or substantial portions of the
+    Software.
+
+    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+    IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+    FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL
+    THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+    LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+    FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+    DEALINGS IN THE SOFTWARE.
+  </copyright>
+
+  <description summary="Protocol for requesting activation of surfaces">
+    The way for a client to pass focus to another toplevel is as follows.
+
+    The client that intends to activate another toplevel uses the
+    xdg_activation_v1.get_activation_token request to get an activation token.
+    This token is then forwarded to the client, which is supposed to activate
+    one of its surfaces, through a separate band of communication.
+
+    One established way of doing this is through the XDG_ACTIVATION_TOKEN
+    environment variable of a newly launched child process. The child process
+    should unset the environment variable again right after reading it out in
+    order to avoid propagating it to other child processes.
+
+    Another established way exists for Applications implementing the D-Bus
+    interface org.freedesktop.Application, which should get their token under
+    activation-token on their platform_data.
+
+    In general activation tokens may be transferred across clients through
+    means not described in this protocol.
+
+    The client to be activated will then pass the token
+    it received to the xdg_activation_v1.activate request. The compositor can
+    then use this token to decide how to react to the activation request.
+
+    The token the activating client gets may be ineffective either already at
+    the time it receives it, for example if it was not focused, for focus
+    stealing prevention. The activating client will have no way to discover
+    the validity of the token, and may still forward it to the to be activated
+    client.
+
+    The created activation token may optionally get information attached to it
+    that can be used by the compositor to identify the application that we
+    intend to activate. This can for example be used to display a visual hint
+    about what application is being started.
+
+    Warning! The protocol described in this file is currently in the testing
+    phase. Backward compatible changes may be added together with the
+    corresponding interface version bump. Backward incompatible changes can
+    only be done by creating a new major version of the extension.
+  </description>
+
+  <interface name="xdg_activation_v1" version="1">
+    <description summary="interface for activating surfaces">
+      A global interface used for informing the compositor about applications
+      being activated or started, or for applications to request to be
+      activated.
+    </description>
+
+    <request name="destroy" type="destructor">
+      <description summary="destroy the xdg_activation object">
+        Notify the compositor that the xdg_activation object will no longer be
+        used.
+
+        The child objects created via this interface are unaffected and should
+        be destroyed separately.
+      </description>
+    </request>
+
+    <request name="get_activation_token">
+      <description summary="requests a token">
+        Creates an xdg_activation_token_v1 object that will provide
+        the initiating client with a unique token for this activation. This
+        token should be offered to the clients to be activated.
+      </description>
+
+      <arg name="id" type="new_id" interface="xdg_activation_token_v1"/>
+    </request>
+
+    <request name="activate">
+      <description summary="notify new interaction being available">
+        Requests surface activation. It's up to the compositor to display
+        this information as desired, for example by placing the surface above
+        the rest.
+
+        The compositor may know who requested this by checking the activation
+        token and might decide not to follow through with the activation if it's
+        considered unwanted.
+
+        Compositors can ignore unknown activation tokens when an invalid
+        token is passed.
+      </description>
+      <arg name="token" type="string" summary="the activation token of the initiating client"/>
+      <arg name="surface" type="object" interface="wl_surface"
+	   summary="the wl_surface to activate"/>
+    </request>
+  </interface>
+
+  <interface name="xdg_activation_token_v1" version="1">
+    <description summary="an exported activation handle">
+      An object for setting up a token and receiving a token handle that can
+      be passed as an activation token to another client.
+
+      The object is created using the xdg_activation_v1.get_activation_token
+      request. This object should then be populated with the app_id, surface
+      and serial information and committed. The compositor shall then issue a
+      done event with the token. In case the request's parameters are invalid,
+      the compositor will provide an invalid token.
+    </description>
+
+    <enum name="error">
+      <entry name="already_used" value="0"
+             summary="The token has already been used previously"/>
+    </enum>
+
+    <request name="set_serial">
+      <description summary="specifies the seat and serial of the activating event">
+        Provides information about the seat and serial event that requested the
+        token.
+
+        The serial can come from an input or focus event. For instance, if a
+        click triggers the launch of a third-party client, the launcher client
+        should send a set_serial request with the serial and seat from the
+        wl_pointer.button event.
+
+        Some compositors might refuse to activate toplevels when the token
+        doesn't have a valid and recent enough event serial.
+
+        Must be sent before commit. This information is optional.
+      </description>
+      <arg name="serial" type="uint"
+           summary="the serial of the event that triggered the activation"/>
+      <arg name="seat" type="object" interface="wl_seat"
+           summary="the wl_seat of the event"/>
+    </request>
+
+    <request name="set_app_id">
+      <description summary="specifies the application being activated">
+        The requesting client can specify an app_id to associate the token
+        being created with it.
+
+        Must be sent before commit. This information is optional.
+      </description>
+      <arg name="app_id" type="string"
+           summary="the application id of the client being activated."/>
+    </request>
+
+    <request name="set_surface">
+      <description summary="specifies the surface requesting activation">
+        This request sets the surface requesting the activation. Note, this is
+        different from the surface that will be activated.
+
+        Some compositors might refuse to activate toplevels when the token
+        doesn't have a requesting surface.
+
+        Must be sent before commit. This information is optional.
+      </description>
+      <arg name="surface" type="object" interface="wl_surface"
+	   summary="the requesting surface"/>
+    </request>
+
+    <request name="commit">
+      <description summary="issues the token request">
+        Requests an activation token based on the different parameters that
+        have been offered through set_serial, set_surface and set_app_id.
+      </description>
+    </request>
+
+    <event name="done">
+      <description summary="the exported activation token">
+        The 'done' event contains the unique token of this activation request
+        and notifies that the provider is done.
+      </description>
+      <arg name="token" type="string" summary="the exported activation token"/>
+    </event>
+
+    <request name="destroy" type="destructor">
+      <description summary="destroy the xdg_activation_token_v1 object">
+        Notify the compositor that the xdg_activation_token_v1 object will no
+        longer be used. The received token stays valid.
+      </description>
+    </request>
+  </interface>
+</protocol>
diff --git a/Sources/Server/OwlServer.m b/Sources/Server/OwlServer.m
index d650cc7..96e2d29 100644
--- a/Sources/Server/OwlServer.m
+++ b/Sources/Server/OwlServer.m
@@ -24,6 +24,7 @@
 #import "OwlWlShell.h"
 #import "OwlZxdgShellV6.h"
 #import "OwlXdgWmBase.h"
+#import "OwlXdgActivationV1.h"
 #import "OwlZxdgDecorationManagerV1.h"
 #import "OwlWpContentTypeManagerV1.h"
 #import "OwlWpSinglePixelBufferManagerV1.h"
@@ -129,6 +130,7 @@
     [OwlWlShell addGlobalToDisplay: _display];
     [OwlZxdgShellV6 addGlobalToDisplay: _display];
     [OwlXdgWmBase addGlobalToDisplay: _display];
+    [OwlXdgActivationV1 addGlobalToDisplay: _display];
     [OwlZxdgDecorationManagerV1 addGlobalToDisplay: _display];
     [OwlWpContentTypeManagerV1 addGlobalToDisplay: _display];
     [OwlWpSinglePixelBufferManagerV1 addGlobalToDisplay: _display];
diff --git a/Sources/Shell/OwlXdgActivationV1.h b/Sources/Shell/OwlXdgActivationV1.h
new file mode 100644
index 0000000..251ab9c
--- /dev/null
+++ b/Sources/Shell/OwlXdgActivationV1.h
@@ -0,0 +1,35 @@
+/* This file is part of Owl.
+ *
+ * Copyright © 2019-2021 Sergey Bugaev <bugaevc@gmail.com>
+ *
+ * Owl is free software: you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published
+ * by the Free Software Foundation, either version 3 of the License,
+ * or (at your option) any later version.
+ *
+ * Owl is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Owl.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#import "OwlGlobal.h"
+#import <Cocoa/Cocoa.h>
+#import <wayland-server.h>
+
+
+/* Implements xdg_activation_v1. Tokens are not validated: activate
+ * accepts any token, which is lenient but fine for a v1 implementation
+ * (foot only uses this for bell.urgent). */
+@interface OwlXdgActivationV1 : NSObject <OwlGlobal> {
+    struct wl_resource *_resource;
+}
+
+- (id) initWithResource: (struct wl_resource *) resource;
+
++ (void) addGlobalToDisplay: (struct wl_display *) display;
+
+@end
diff --git a/Sources/Shell/OwlXdgActivationV1.m b/Sources/Shell/OwlXdgActivationV1.m
new file mode 100644
index 0000000..eb3786c
--- /dev/null
+++ b/Sources/Shell/OwlXdgActivationV1.m
@@ -0,0 +1,211 @@
+/* This file is part of Owl.
+ *
+ * Copyright © 2019-2021 Sergey Bugaev <bugaevc@gmail.com>
+ *
+ * Owl is free software: you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published
+ * by the Free Software Foundation, either version 3 of the License,
+ * or (at your option) any later version.
+ *
+ * Owl is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Owl.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#import "OwlXdgActivationV1.h"
+#import "OwlSurface.h"
+#import "OwlXdgToplevel.h"
+#import "OwlWindowWrapper.h"
+#import "OwlWindow.h"
+#import "xdg-activation-v1.h"
+#import <wayland-server.h>
+
+
+@interface OwlXdgActivationTokenV1 : NSObject {
+@public
+    struct wl_resource *_resource;
+}
+
+- (id) initWithResource: (struct wl_resource *) resource;
+
+@end
+
+@implementation OwlXdgActivationTokenV1
+
+static void activation_token_destroy(struct wl_resource *resource) {
+    OwlXdgActivationTokenV1 *self = wl_resource_get_user_data(resource);
+    [self release];
+}
+
+static void activation_token_destroy_handler(
+    struct wl_client *client,
+    struct wl_resource *resource
+) {
+    wl_resource_destroy(resource);
+}
+
+static void activation_token_set_serial_handler(
+    struct wl_client *client,
+    struct wl_resource *resource,
+    uint32_t serial,
+    struct wl_resource *seat_resource
+) {
+    // Not tracked: tokens are not validated on activate.
+}
+
+static void activation_token_set_app_id_handler(
+    struct wl_client *client,
+    struct wl_resource *resource,
+    const char *app_id
+) {
+    // Not tracked: tokens are not validated on activate.
+}
+
+static void activation_token_set_surface_handler(
+    struct wl_client *client,
+    struct wl_resource *resource,
+    struct wl_resource *surface_resource
+) {
+    // Not tracked: tokens are not validated on activate.
+}
+
+static void activation_token_commit_handler(
+    struct wl_client *client,
+    struct wl_resource *resource
+) {
+    static unsigned counter;
+    char token[32];
+    snprintf(token, sizeof(token), "owl-%u", ++counter);
+    xdg_activation_token_v1_send_done(resource, token);
+}
+
+static const struct xdg_activation_token_v1_interface
+activation_token_impl = {
+    .set_serial = activation_token_set_serial_handler,
+    .set_app_id = activation_token_set_app_id_handler,
+    .set_surface = activation_token_set_surface_handler,
+    .commit = activation_token_commit_handler,
+    .destroy = activation_token_destroy_handler
+};
+
+- (id) initWithResource: (struct wl_resource *) resource {
+    _resource = resource;
+    wl_resource_set_implementation(
+        resource,
+        &activation_token_impl,
+        [self retain],
+        activation_token_destroy
+    );
+    return self;
+}
+
+@end
+
+
+@implementation OwlXdgActivationV1
+
+static void activation_destroy_handler(
+    struct wl_client *client,
+    struct wl_resource *resource
+) {
+    wl_resource_destroy(resource);
+}
+
+static void activation_get_activation_token_handler(
+    struct wl_client *client,
+    struct wl_resource *resource,
+    uint32_t id
+) {
+    struct wl_resource *token_resource = wl_resource_create(
+        client,
+        &xdg_activation_token_v1_interface,
+        wl_resource_get_version(resource),
+        id
+    );
+    [[[OwlXdgActivationTokenV1 alloc] initWithResource: token_resource] release];
+}
+
+static void activation_activate_handler(
+    struct wl_client *client,
+    struct wl_resource *resource,
+    const char *token,
+    struct wl_resource *surface_resource
+) {
+    // No token validation store: any token (including none we
+    // issued) is accepted, matching the lenient v1 behavior noted
+    // in the brief.
+    OwlSurface *surface = wl_resource_get_user_data(surface_resource);
+    if (surface == nil) {
+        return;
+    }
+    id<OwlSurfaceRole> role = [surface role];
+    if (![role isKindOfClass: [OwlXdgToplevel class]]) {
+        return;
+    }
+    OwlWindowWrapper *wrapper = [(OwlXdgToplevel *) role windowWrapper];
+    if (wrapper == nil) {
+        return;
+    }
+    OwlWindow *window = [wrapper window];
+    if (window == nil) {
+        return;
+    }
+
+    [window makeKeyAndOrderFront: nil];
+    if (![NSApp isActive]) {
+        [NSApp requestUserAttention: NSInformationalRequest];
+    }
+}
+
+static const struct xdg_activation_v1_interface activation_impl = {
+    .destroy = activation_destroy_handler,
+    .get_activation_token = activation_get_activation_token_handler,
+    .activate = activation_activate_handler
+};
+
+static void activation_destroy(struct wl_resource *resource) {
+    OwlXdgActivationV1 *self = wl_resource_get_user_data(resource);
+    [self release];
+}
+
+- (id) initWithResource: (struct wl_resource *) resource {
+    _resource = resource;
+    wl_resource_set_implementation(
+        resource,
+        &activation_impl,
+        [self retain],
+        activation_destroy
+    );
+    return self;
+}
+
+static void activation_bind(
+    struct wl_client *client,
+    void *data,
+    uint32_t version,
+    uint32_t id
+) {
+    struct wl_resource *resource = wl_resource_create(
+        client,
+        &xdg_activation_v1_interface,
+        version,
+        id
+    );
+    [[[OwlXdgActivationV1 alloc] initWithResource: resource] release];
+}
+
++ (void) addGlobalToDisplay: (struct wl_display *) display {
+    wl_global_create(
+        display,
+        &xdg_activation_v1_interface,
+        1,
+        NULL,
+        activation_bind
+    );
+}
+
+@end
diff --git a/Sources/Shell/OwlXdgToplevel.h b/Sources/Shell/OwlXdgToplevel.h
index 41f49ee..4316f74 100644
--- a/Sources/Shell/OwlXdgToplevel.h
+++ b/Sources/Shell/OwlXdgToplevel.h
@@ -39,4 +39,6 @@
 
 - (void) sendConfigureWithSize: (NSSize) size;
 
+- (OwlWindowWrapper *) windowWrapper;
+
 @end
diff --git a/Sources/Shell/OwlXdgToplevel.m b/Sources/Shell/OwlXdgToplevel.m
index 7a1629a..ee4d6ef 100644
--- a/Sources/Shell/OwlXdgToplevel.m
+++ b/Sources/Shell/OwlXdgToplevel.m
@@ -251,6 +251,10 @@ static const struct xdg_toplevel_interface xdg_toplevel_impl = {
     wl_client_flush(client);
 }
 
+- (OwlWindowWrapper *) windowWrapper {
+    return _window;
+}
+
 - (OwlKeyboard *) keyboard {
     struct wl_client *client = wl_resource_get_client(_resource);
     return [OwlKeyboard keyboardForClient: client];
-- 
2.43.0

